### ISSUE TITLE
np.full and object arrays

### ISSUE THREAD

**Issue by anntzer** [numpy/numpy#6366]
`np.full` says that the resulting array's dtype is `np.array(fill_value).dtype` without mention of object arrays, but
```
In [1]: np.full((1,), None)
Out[1]: array([ nan])
```
and
```
In [2]: class C: pass

In [3]: np.full((1,), C())
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-e364c2e90930> in <module>()
----> 1 np.full((1,), C())

/usr/lib/python3.4/site-packages/numpy/core/numeric.py in full(shape, fill_value, dtype, order)
    286     """
    287     a = empty(shape, dtype, order)
--> 288     multiarray.copyto(a, fill_value, casting='unsafe')
    289     return a
    290 

TypeError: float() argument must be a string or a number, not 'C'
```

**Comment by seberg** [comment]
Hah, the code says the exact opposite. That is too bad, but I think we should probably change the documentation?

**Comment by anntzer** [comment]
Indeed, I just realized that the problem is not just with object arrays.  Ugh.

**Comment by seberg** [comment]
I am not sure whether we have the choice to change the default to what the documentation says. On first sight the documentations idea does not seem unreasonable at all, but it does change behaviour (most importantly, return integer instead of float array).

**Comment by anntzer** [comment]
Weirdly enough, this point of the docs is explicitly discussed in #2875 (the original PR for `full`).  I see however that the tests fail to cover the case of `full` (as well as `empty`, `zeros` and `ones`) being called with no `dtype` argument.
I'd rather fix the current behavior, which is clearly wrong IMO, possibly going through some deprecation warning period first?

**Comment by njsmith** [comment]
Can you send a note to the list explaining the current and desired behavior
and purpose to change it as a bug fix? It's a very new function so there
probably aren't many people relying on it, but it's good to double check.

**Comment by anntzer** [comment]
Have you received https://mail.scipy.org/pipermail/numpy-discussion/2015-September/073790.html?  Somehow the email didn't get to my inbox, so I wanted to make sure it didn't get lost somehow.  (Possibly just some stupid gmail setting.)

**Comment by njsmith** [comment]
I did receive it (and just replied), but cc: @infowolfe just in case it's related to the other problems the mailing list has been having.

**Comment by anntzer** [comment]
Bumping this: can we switch to the new behavior in 1.11?

**Comment by njsmith** [comment]
So it's been working this way since it was introduced in 1.8, back in Oct
2013? Ugh. I agree that the current behavior is wrong and self-defeating,
but that's a bit long to just break without warning. I vote we do a
FutureWarning in 1.11 (if @charris doesn't object) and switch the behavior
after an accelerated-but-not-instantaneous deprecation period. How does
that sound?

On Tue, Jan 19, 2016 at 10:57 AM, Antony Lee <notifications@github.com>
wrote:

> Bumping this: can we switch to the new behavior in 1.11?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/issues/6366#issuecomment-172951154>.
>



-- 
Nathaniel J. Smith -- https://vorpus.org <http://vorpus.org>

**Comment by seberg** [comment]
@njsmith, the FutureWarning is already there for 1.10 I think, so we could think about changing already, considering how new the function was back then.

**Comment by njsmith** [comment]
Ah, so it is. 1.11 still seems a bit early to make the change, given that
1.10 is still propagating out to distros etc.
.
+1 to making the change on master for 1.12, though

On Tue, Jan 19, 2016 at 12:03 PM, seberg <notifications@github.com> wrote:

> @njsmith <https://github.com/njsmith>, the FutureWarning is already there
> for 1.10 I think, so we could think about changing already, considering how
> new the function was back then.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/issues/6366#issuecomment-172969933>.
>



-- 
Nathaniel J. Smith -- https://vorpus.org <http://vorpus.org>

**Comment by anntzer** [comment]
Essentially handled by #7437, except that the behavior for object dtypes when the fill value is an iterable itself is questionable.  See discussion in #7437.

**Comment by eric-wieser** [comment]
@anntzer: Can this be closed in favor of #5972 in your opinion?

**Comment by anntzer** [comment]
Sure.
